File: packages/core/src/shapes/symexpr.ts
/**
 * Exact symbolic arithmetic for tensor dimensions.
 *
 * A `Sym` is a multivariate polynomial with rational coefficients over named
 * symbols (`B`, `T`, `D`, `H`, `dh`, ...). This is exactly the expressive power
 * tensor shapes need: products for reshapes (`(H dh)`), sums for concatenation,
 * and exact divisibility obligations for splits.
 *
 * Design symbols (D, H, F, ...) always have a concrete numeric value; runtime
 * symbols (B, T) stay indeterminate. Two shapes are compatible when their
 * difference is the zero polynomial after substituting the concrete values.
 */

// ---------------------------------------------------------------------------
// Rational numbers
// ---------------------------------------------------------------------------

export interface Rat {
  readonly n: number;
  readonly d: number;
}

function igcd(a: number, b: number): number {
  a = Math.abs(a);
  b = Math.abs(b);
  while (b) {
    const t = a % b;
    a = b;
    b = t;
  }
  return a || 1;
}

/** Build a normalized rational. Accepts non-integer inputs (e.g. 1.3) exactly. */
export function rat(n: number, d = 1): Rat {
  if (d === 0) throw new Error("SymExpr: division by zero");
  if (!Number.isFinite(n) || !Number.isFinite(d)) throw new Error("SymExpr: non-finite value");
  if (!Number.isInteger(n) || !Number.isInteger(d)) {
    let scale = 1;
    while ((!Number.isInteger(n * scale) || !Number.isInteger(d * scale)) && scale < 1e12) scale *= 10;
    n = Math.round(n * scale);
    d = Math.round(d * scale);
  }
  if (d < 0) {
    n = -n;
    d = -d;
  }
  const g = igcd(n, d);
  return { n: n / g, d: d / g };
}

const RAT_ZERO: Rat = { n: 0, d: 1 };
const RAT_ONE: Rat = { n: 1, d: 1 };

const ratAdd = (a: Rat, b: Rat): Rat => rat(a.n * b.d + b.n * a.d, a.d * b.d);
const ratMul = (a: Rat, b: Rat): Rat => rat(a.n * b.n, a.d * b.d);
const ratDiv = (a: Rat, b: Rat): Rat => {
  if (b.n === 0) throw new Error("SymExpr: division by zero");
  return rat(a.n * b.d, a.d * b.n);
};
const ratNeg = (a: Rat): Rat => ({ n: -a.n, d: a.d });
const ratIsZero = (a: Rat): boolean => a.n === 0;
const ratIsOne = (a: Rat): boolean => a.n === 1 && a.d === 1;
const ratNum = (a: Rat): number => a.n / a.d;

function ratToString(a: Rat): string {
  return a.d === 1 ? String(a.n) : `${a.n}/${a.d}`;
}

// ---------------------------------------------------------------------------
// Monomials
// ---------------------------------------------------------------------------

/** Exponent per symbol. Absent symbol means exponent 0. */
export type Mono = Readonly<Record<string, number>>;

function monoKey(m: Mono): string {
  const names = Object.keys(m)
    .filter((k) => m[k] !== 0)
    .sort();
  if (names.length === 0) return "";
  return names.map((k) => (m[k] === 1 ? k : `${k}^${m[k]}`)).join("*");
}

function monoMul(a: Mono, b: Mono): Mono {
  const out: Record<string, number> = { ...a };
  for (const k of Object.keys(b)) {
    const e = (out[k] ?? 0) + b[k];
    if (e === 0) delete out[k];
    else out[k] = e;
  }
  return out;
}

function monoDegree(m: Mono): number {
  let d = 0;
  for (const k of Object.keys(m)) d += m[k];
  return d;
}

interface Term {
  readonly c: Rat;
  readonly m: Mono;
}

// ---------------------------------------------------------------------------
// Sym
// ---------------------------------------------------------------------------

export class Sym {
  /** Canonical monomial key -> term. Never contains zero coefficients. */
  private readonly terms: ReadonlyMap<string, Term>;

  private constructor(terms: Map<string, Term>) {
    this.terms = terms;
  }

  // --- construction -------------------------------------------------------

  static readonly ZERO = new Sym(new Map());

  static con(v: number | Rat): Sym {
    const r = typeof v === "number" ? rat(v) : v;
    if (ratIsZero(r)) return Sym.ZERO;
    return new Sym(new Map([["", { c: r, m: {} }]]));
  }

  static v(name: string): Sym {
    return new Sym(new Map([[name, { c: RAT_ONE, m: { [name]: 1 } }]]));
  }

  static sum(parts: Sym[]): Sym {
    return parts.reduce<Sym>((a, b) => a.add(b), Sym.ZERO);
  }

  static product(parts: Sym[]): Sym {
    return parts.reduce<Sym>((a, b) => a.mul(b), Sym.con(1));
  }

  // --- arithmetic ---------------------------------------------------------

  add(other: Sym): Sym {
    const out = new Map<string, Term>(this.terms);
    for (const [k, t] of other.terms) {
      const cur = out.get(k);
      if (!cur) {
        out.set(k, t);
      } else {
        const c = ratAdd(cur.c, t.c);
        if (ratIsZero(c)) out.delete(k);
        else out.set(k, { c, m: cur.m });
      }
    }
    return new Sym(out);
  }

  neg(): Sym {
    const out = new Map<string, Term>();
    for (const [k, t] of this.terms) out.set(k, { c: ratNeg(t.c), m: t.m });
    return new Sym(out);
  }

  sub(other: Sym): Sym {
    return this.add(other.neg());
  }

  mul(other: Sym): Sym {
    const out = new Map<string, Term>();
    for (const a of this.terms.values()) {
      for (const b of other.terms.values()) {
        const m = monoMul(a.m, b.m);
        const k = monoKey(m);
        const c = ratMul(a.c, b.c);
        const cur = out.get(k);
        if (!cur) {
          out.set(k, { c, m });
        } else {
          const nc = ratAdd(cur.c, c);
          if (ratIsZero(nc)) out.delete(k);
          else out.set(k, { c: nc, m });
        }
      }
    }
    return new Sym(out);
  }

  pow(k: number): Sym {
    if (!Number.isInteger(k) || k < 0) throw new Error(`SymExpr: unsupported exponent ${k}`);
    let acc = Sym.con(1);
    for (let i = 0; i < k; i++) acc = acc.mul(this);
    return acc;
  }

  /**
   * Exact division. Succeeds when `other` is a single term (constant or
   * monomial) that divides every term of this expression, or when the result is
   * otherwise provably exact. Returns null when divisibility cannot be proven,
   * which callers turn into a design-rule obligation.
   */
  divExact(other: Sym): Sym | null {
    if (other.isZero()) return null;
    if (other.terms.size !== 1) {
      // Only handle the case where this is an exact multiple by trying the
      // constant-quotient shortcut.
      const oc = other.asConst();
      if (oc === null) return null;
      return this.divExact(Sym.con(oc));
    }
    const [b] = [...other.terms.values()];
    const out = new Map<string, Term>();
    for (const a of this.terms.values()) {
      const m: Record<string, number> = { ...a.m };
      for (const k of Object.keys(b.m)) {
        const e = (m[k] ?? 0) - b.m[k];
        if (e < 0) return null;
        if (e === 0) delete m[k];
        else m[k] = e;
      }
      const c = ratDiv(a.c, b.c);
      out.set(monoKey(m), { c, m });
    }
    return new Sym(out);
  }

  // --- inspection ---------------------------------------------------------

  isZero(): boolean {
    return this.terms.size === 0;
  }

  /** Numeric value when this is a constant, else null. */
  asConst(): number | null {
    if (this.terms.size === 0) return 0;
    if (this.terms.size === 1) {
      const t = this.terms.get("");
      if (t) return ratNum(t.c);
    }
    return null;
  }

  symbols(): string[] {
    const s = new Set<string>();
    for (const t of this.terms.values()) for (const k of Object.keys(t.m)) s.add(k);
    return [...s].sort();
  }

  /** True when this is exactly a single named symbol with coefficient 1. */
  asAtom(): string | null {
    if (this.terms.size !== 1) return null;
    const [t] = [...this.terms.values()];
    if (!ratIsOne(t.c)) return null;
    const names = Object.keys(t.m);
    if (names.length !== 1 || t.m[names[0]] !== 1) return null;
    return names[0];
  }

  /** Substitute a subset of symbols with numbers, leaving the rest symbolic. */
  subst(values: Readonly<Record<string, number>>): Sym {
    let acc = Sym.ZERO;
    for (const t of this.terms.values()) {
      let term = Sym.con(t.c);
      for (const k of Object.keys(t.m)) {
        const e = t.m[k];
        const base = k in values ? Sym.con(values[k]) : Sym.v(k);
        term = term.mul(base.pow(e));
      }
      acc = acc.add(term);
    }
    return acc;
  }

  /** Fully evaluate; returns null when some symbol has no value. */
  toNumber(values: Readonly<Record<string, number>>): number | null {
    return this.subst(values).asConst();
  }

  /** Structural equality of the canonical form (no substitution). */
  equals(other: Sym): boolean {
    return this.sub(other).isZero();
  }

  /**
   * Equality under a partial environment. Used for shape compatibility: design
   * symbols are substituted, runtime symbols stay indeterminate, and the
   * difference must be the zero polynomial.
   */
  equalsUnder(other: Sym, values: Readonly<Record<string, number>>): boolean {
    return this.sub(other).subst(values).isZero();
  }

  toString(): string {
    if (this.terms.size === 0) return "0";
    const list = [...this.terms.values()].sort((a, b) => {
      const da = monoDegree(a.m);
      const db = monoDegree(b.m);
      if (da !== db) return db - da;
      return monoKey(a.m).localeCompare(monoKey(b.m));
    });
    let out = "";
    for (let i = 0; i < list.length; i++) {
      const t = list[i];
      const negative = t.c.n < 0;
      const abs: Rat = negative ? ratNeg(t.c) : t.c;
      if (i === 0) out += negative ? "-" : "";
      else out += negative ? " - " : " + ";
      const key = monoKey(t.m);
      if (key === "") {
        out += ratToString(abs);
      } else if (ratIsOne(abs)) {
        out += key;
      } else {
        out += `${ratToString(abs)}*${key}`;
      }
    }
    return out;
  }
}

/** Convenience: a constant-zero check that tolerates null. */
export function symEq(a: Sym | null, b: Sym | null, values: Readonly<Record<string, number>>): boolean {
  if (!a || !b) return false;
  return a.equalsUnder(b, values);
}

File: packages/core-go/shapes/symexpr.go
// Exact symbolic arithmetic for tensor dimensions.
//
// A Sym is a multivariate polynomial with rational coefficients over named
// symbols (B, T, D, H, dh, ...). This is exactly the expressive power tensor
// shapes need: products for reshapes, sums for concatenation, and exact
// divisibility obligations for splits.
//
// Design symbols (D, H, F, ...) always have a concrete numeric value; runtime
// symbols (B, T) stay indeterminate. Two shapes are compatible when their
// difference is the zero polynomial after substituting the concrete values.
package shapes

import (
	"fmt"
	"math"
	"math/big"
	"sort"
	"strings"
)

// ---------------------------------------------------------------------------
// Rational numbers
// ---------------------------------------------------------------------------

var (
	ratOne = new(big.Rat).SetInt64(1)
)

// RatFrom builds a rational from a float.
//
// Decimal literals have to come out as the decimal the author wrote: RatFrom of
// 1.3 is 13/10, not the binary 5854679515581645/4503599627370496 that
// SetFloat64 would give, because 1.3*8/3*D is how a feed-forward width is
// written and 13/10 is what it means. The scaling loop is the TypeScript
// engine's, kept so the two agree term for term.
//
// Numerator and denominator are arbitrary precision here rather than the
// doubles the TypeScript carried, which are exact to 2^53 and silently wrong
// past it. A 671-billion-parameter model is comfortably inside that; its FLOP
// counts are not.
func RatFrom(n float64) *big.Rat { return RatFromPair(n, 1) }

// RatFromPair builds n/d, scaling both until they are whole.
//
// It is the TypeScript engine's `rat()`, kept step for step rather than
// improved. Scaling by ten until both sides are integral is not the same as
// exact conversion — 1/3 comes out as 333333333333/1000000000000 — and making
// it exact here would have the two engines print different shapes for the same
// design. Faithful first; the imprecision can be fixed in both at once.
func RatFromPair(n, d float64) *big.Rat {
	if d == 0 {
		panic("shapes: division by zero")
	}
	if math.IsInf(n, 0) || math.IsNaN(n) || math.IsInf(d, 0) || math.IsNaN(d) {
		panic("shapes: non-finite value")
	}
	if isWhole(n) && isWhole(d) && math.Abs(n) < 1e18 && math.Abs(d) < 1e18 {
		return new(big.Rat).SetFrac64(int64(n), int64(d))
	}
	scale := 1.0
	for (!isWhole(n*scale) || !isWhole(d*scale)) && scale < 1e12 {
		scale *= 10
	}
	return new(big.Rat).SetFrac64(int64(math.Round(n*scale)), int64(math.Round(d*scale)))
}

func isWhole(v float64) bool { return v == math.Trunc(v) && !math.IsInf(v, 0) }

// RatFromDecimal parses a decimal literal exactly, so "1.3" gives 13/10.
func RatFromDecimal(text string) (*big.Rat, error) {
	r, ok := new(big.Rat).SetString(text)
	if !ok {
		return nil, fmt.Errorf("bad number %q", text)
	}
	return r, nil
}

func ratIsZero(a *big.Rat) bool { return a.Sign() == 0 }
func ratIsOne(a *big.Rat) bool  { return a.Cmp(ratOne) == 0 }

func ratString(a *big.Rat) string {
	if a.IsInt() {
		return a.Num().String()
	}
	return a.Num().String() + "/" + a.Denom().String()
}

// ---------------------------------------------------------------------------
// Monomials
// ---------------------------------------------------------------------------

// Mono is the exponent per symbol. An absent symbol means exponent 0.
type Mono map[string]int

func monoKey(m Mono) string {
	names := make([]string, 0, len(m))
	for k, e := range m {
		if e != 0 {
			names = append(names, k)
		}
	}
	if len(names) == 0 {
		return ""
	}
	sort.Strings(names)
	parts := make([]string, len(names))
	for i, k := range names {
		if m[k] == 1 {
			parts[i] = k
		} else {
			parts[i] = fmt.Sprintf("%s^%d", k, m[k])
		}
	}
	return strings.Join(parts, "*")
}

func monoMul(a, b Mono) Mono {
	out := make(Mono, len(a)+len(b))
	for k, e := range a {
		out[k] = e
	}
	for k, e := range b {
		if n := out[k] + e; n == 0 {
			delete(out, k)
		} else {
			out[k] = n
		}
	}
	return out
}

func monoDegree(m Mono) int {
	d := 0
	for _, e := range m {
		d += e
	}
	return d
}

type term struct {
	c *big.Rat
	m Mono
}

// ---------------------------------------------------------------------------
// Sym
// ---------------------------------------------------------------------------

// Sym is a multivariate polynomial in canonical form: one term per distinct
// monomial, never with a zero coefficient. The zero value is the zero
// polynomial, so a declared Sym is usable without construction.
//
// Every method treats the receiver as immutable and returns a fresh value.
type Sym struct {
	terms map[string]term
}

// Zero is the zero polynomial.
func Zero() Sym { return Sym{} }

// Con is a constant.
func Con(v float64) Sym { return ConRat(RatFrom(v)) }

// ConRat is a constant from an exact rational.
func ConRat(r *big.Rat) Sym {
	if ratIsZero(r) {
		return Sym{}
	}
	return Sym{terms: map[string]term{"": {c: new(big.Rat).Set(r), m: Mono{}}}}
}

// V is a single named symbol.
func V(name string) Sym {
	return Sym{terms: map[string]term{name: {c: new(big.Rat).Set(ratOne), m: Mono{name: 1}}}}
}

// Sum adds every part.
func Sum(parts ...Sym) Sym {
	acc := Zero()
	for _, p := range parts {
		acc = acc.Add(p)
	}
	return acc
}

// Product multiplies every part.
func Product(parts ...Sym) Sym {
	acc := Con(1)
	for _, p := range parts {
		acc = acc.Mul(p)
	}
	return acc
}

func (s Sym) clone() map[string]term {
	out := make(map[string]term, len(s.terms))
	for k, t := range s.terms {
		out[k] = t
	}
	return out
}

// Add returns s + other.
func (s Sym) Add(other Sym) Sym {
	out := s.clone()
	for k, t := range other.terms {
		cur, ok := out[k]
		if !ok {
			out[k] = t
			continue
		}
		c := new(big.Rat).Add(cur.c, t.c)
		if ratIsZero(c) {
			delete(out, k)
		} else {
			out[k] = term{c: c, m: cur.m}
		}
	}
	return Sym{terms: out}
}

// Neg returns -s.
func (s Sym) Neg() Sym {
	out := make(map[string]term, len(s.terms))
	for k, t := range s.terms {
		out[k] = term{c: new(big.Rat).Neg(t.c), m: t.m}
	}
	return Sym{terms: out}
}

// Sub returns s - other.
func (s Sym) Sub(other Sym) Sym { return s.Add(other.Neg()) }

// Mul returns s * other.
func (s Sym) Mul(other Sym) Sym {
	out := map[string]term{}
	for _, a := range s.terms {
		for _, b := range other.terms {
			m := monoMul(a.m, b.m)
			k := monoKey(m)
			c := new(big.Rat).Mul(a.c, b.c)
			cur, ok := out[k]
			if !ok {
				out[k] = term{c: c, m: m}
				continue
			}
			nc := new(big.Rat).Add(cur.c, c)
			if ratIsZero(nc) {
				delete(out, k)
			} else {
				out[k] = term{c: nc, m: m}
			}
		}
	}
	return Sym{terms: out}
}

// Pow raises s to a non-negative integer power.
func (s Sym) Pow(k int) (Sym, error) {
	if k < 0 {
		return Zero(), fmt.Errorf("shapes: unsupported exponent %d", k)
	}
	acc := Con(1)
	for i := 0; i < k; i++ {
		acc = acc.Mul(s)
	}
	return acc, nil
}

// DivExact divides exactly.
//
// It succeeds when other is a single term — a constant or a monomial — that
// divides every term of s. It reports ok=false when divisibility cannot be
// proven, which callers turn into a design-rule obligation rather than an
// approximation.
func (s Sym) DivExact(other Sym) (Sym, bool) {
	if other.IsZero() {
		return Zero(), false
	}
	if len(other.terms) != 1 {
		if oc, ok := other.AsConst(); ok {
			return s.DivExact(Con(oc))
		}
		return Zero(), false
	}
	var b term
	for _, t := range other.terms {
		b = t
	}
	out := map[string]term{}
	for _, a := range s.terms {
		m := make(Mono, len(a.m))
		for k, e := range a.m {
			m[k] = e
		}
		for k, e := range b.m {
			n := m[k] - e
			if n < 0 {
				return Zero(), false
			}
			if n == 0 {
				delete(m, k)
			} else {
				m[k] = n
			}
		}
		out[monoKey(m)] = term{c: new(big.Rat).Quo(a.c, b.c), m: m}
	}
	return Sym{terms: out}, true
}

// IsZero reports whether this is the zero polynomial.
func (s Sym) IsZero() bool { return len(s.terms) == 0 }

// AsConst returns the numeric value when s is a constant.
func (s Sym) AsConst() (float64, bool) {
	if len(s.terms) == 0 {
		return 0, true
	}
	if len(s.terms) == 1 {
		if t, ok := s.terms[""]; ok {
			f, _ := t.c.Float64()
			return f, true
		}
	}
	return 0, false
}

// AsExact returns the exact rational value when s is a constant.
func (s Sym) AsExact() (*big.Rat, bool) {
	if len(s.terms) == 0 {
		return new(big.Rat), true
	}
	if len(s.terms) == 1 {
		if t, ok := s.terms[""]; ok {
			return new(big.Rat).Set(t.c), true
		}
	}
	return nil, false
}

// Symbols lists every symbol appearing in s, sorted.
func (s Sym) Symbols() []string {
	seen := map[string]bool{}
	for _, t := range s.terms {
		for k := range t.m {
			seen[k] = true
		}
	}
	out := make([]string, 0, len(seen))
	for k := range seen {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// AsAtom returns the name when s is exactly one named symbol with coefficient 1.
func (s Sym) AsAtom() (string, bool) {
	if len(s.terms) != 1 {
		return "", false
	}
	var t term
	for _, v := range s.terms {
		t = v
	}
	if !ratIsOne(t.c) || len(t.m) != 1 {
		return "", false
	}
	for k, e := range t.m {
		if e != 1 {
			return "", false
		}
		return k, true
	}
	return "", false
}

// Subst replaces a subset of symbols with numbers, leaving the rest symbolic.
func (s Sym) Subst(values map[string]float64) Sym {
	acc := Zero()
	for _, t := range s.terms {
		part := ConRat(t.c)
		for k, e := range t.m {
			var base Sym
			if v, ok := values[k]; ok {
				base = Con(v)
			} else {
				base = V(k)
			}
			p, err := base.Pow(e)
			if err != nil {
				continue
			}
			part = part.Mul(p)
		}
		acc = acc.Add(part)
	}
	return acc
}

// ToNumber fully evaluates; ok is false when some symbol has no value.
func (s Sym) ToNumber(values map[string]float64) (float64, bool) {
	return s.Subst(values).AsConst()
}

// Equals is structural equality of the canonical form, with no substitution.
func (s Sym) Equals(other Sym) bool { return s.Sub(other).IsZero() }

// EqualsUnder is equality under a partial environment.
//
// This is what shape compatibility means: design symbols are substituted,
// runtime symbols stay indeterminate, and the difference has to be the zero
// polynomial. A mismatch is therefore a real polynomial difference rather than
// two numbers that happened not to match.
func (s Sym) EqualsUnder(other Sym, values map[string]float64) bool {
	return s.Sub(other).Subst(values).IsZero()
}

func (s Sym) String() string {
	if len(s.terms) == 0 {
		return "0"
	}
	list := make([]term, 0, len(s.terms))
	for _, t := range s.terms {
		list = append(list, t)
	}
	sort.Slice(list, func(i, j int) bool {
		di, dj := monoDegree(list[i].m), monoDegree(list[j].m)
		if di != dj {
			return di > dj
		}
		return monoKey(list[i].m) < monoKey(list[j].m)
	})
	var b strings.Builder
	for i, t := range list {
		negative := t.c.Sign() < 0
		abs := new(big.Rat).Abs(t.c)
		switch {
		case i == 0 && negative:
			b.WriteString("-")
		case i > 0 && negative:
			b.WriteString(" - ")
		case i > 0:
			b.WriteString(" + ")
		}
		key := monoKey(t.m)
		switch {
		case key == "":
			b.WriteString(ratString(abs))
		case ratIsOne(abs):
			b.WriteString(key)
		default:
			b.WriteString(ratString(abs) + "*" + key)
		}
	}
	return b.String()
}
